[mlir][spirv] Add folding for SelectOp

inbelic opened the pull request:
Add missing constant propogation folder for spirv.Select

Implement additional folding when both selections are equivalent or the condition is a constant Scalar/SplatVector.

Allows for constant folding in the IndexToSPIRV pass.

Part of work #70704

inbelic committed:
[mlir][spirv] Add folding for SelectOp

Add missing constant propogation folder for spirv.Select

Implement additional folding when both selections are equivalent or the
condition is a constant Scalar/SplatVector.

Allows for constant folding in the IndexToSPIRV pass.

Part of work #70704

inbelic commented:
This is a continuation of the contributions that I had started in the fall.

Looking at some lowering paths to SPIR-V now and seeing if there is extra folding that could be done. The SelectOp is required for some folding in MemRef.store and for some IndexToSPIRV lowerings.

kuhar commented on mlir/lib/Dialect/SPIRV/IR/SPIRVCanonicalization.cpp:
```diff
@@ -797,6 +797,85 @@ OpFoldResult spirv::LogicalOrOp::fold(FoldAdaptor adaptor) {
+}
+
+static Attribute foldSelectOp(llvm::ArrayRef<Attribute> operands) {
+  auto condAttrs = llvm::dyn_cast<ElementsAttr>(operands[0]);
+  auto trueAttrs = llvm::dyn_cast<ElementsAttr>(operands[1]);
+  auto falseAttrs = llvm::dyn_cast<ElementsAttr>(operands[2]);
```
nit
```suggestion
  auto condAttrs = dyn_cast<ElementsAttr>(operands[0]);
  auto trueAttrs = dyn_cast<ElementsAttr>(operands[1]);
  auto falseAttrs = dyn_cast<ElementsAttr>(operands[2]);
```

kuhar commented on mlir/lib/Dialect/SPIRV/IR/SPIRVCanonicalization.cpp:
```diff
@@ -797,6 +797,85 @@ OpFoldResult spirv::LogicalOrOp::fold(FoldAdaptor adaptor) {
+
+OpFoldResult spirv::SelectOp::fold(FoldAdaptor adaptor) {
+  // spirv.Select _ x x -> x
+  auto trueVals = getOperand(1);
+  auto falseVals = getOperand(2);
```
Only use auto when the type is obvious based on the RHS. Also in the code below

kuhar commented on mlir/lib/Dialect/SPIRV/IR/SPIRVCanonicalization.cpp:
```diff
@@ -797,6 +797,85 @@ OpFoldResult spirv::LogicalOrOp::fold(FoldAdaptor adaptor) {
+
+OpFoldResult spirv::SelectOp::fold(FoldAdaptor adaptor) {
+  // spirv.Select _ x x -> x
+  auto trueVals = getOperand(1);
+  auto falseVals = getOperand(2);
```
I think this should have named getters like `getCondition()`, `getTrueValue()`, no?
https://github.com/llvm/llvm-project/blob/65ae09eeb6773b14189fc67051870c8fc4eb9ae3/mlir/include/mlir/Dialect/SPIRV/IR/SPIRVLogicalOps.td#L787-L789

inbelic replied:
Yep, much nicer for readability

kuhar commented on mlir/lib/Dialect/SPIRV/IR/SPIRVCanonicalization.cpp:
```diff
@@ -797,6 +797,85 @@ OpFoldResult spirv::LogicalOrOp::fold(FoldAdaptor adaptor) {
+      !condAttrs.getElementType().isInteger(1))
+    return Attribute();
+
+  if (llvm::isa<IntegerType>(elementType)) {
+    return foldSelections<IntegerAttr>(condAttrs, trueAttrs, falseAttrs);
+  } else if (llvm::isa<FloatType>(elementType)) {
+    return foldSelections<FloatAttr>(condAttrs, trueAttrs, falseAttrs);
+  }
```
I don't think we need to distinguish these two cases -- we can construct `DenseElementsAttribute` using the base class `Attribute`

kuhar commented on mlir/lib/Dialect/SPIRV/IR/SPIRVCanonicalization.cpp:
```diff
@@ -797,6 +797,85 @@ OpFoldResult spirv::LogicalOrOp::fold(FoldAdaptor adaptor) {
+  // and Object 2: When a component of Condition is true, the corresponding
+  // component in the result is taken from Object 1, otherwise it is taken
+  // from Object 2.
+  auto elementType = trueAttrs.getElementType();
+  if (trueAttrs.getType() != falseAttrs.getType() ||
+      !condAttrs.getElementType().isInteger(1))
+    return Attribute();
```
This should be ensured by the op verifier, we can drop this check

kuhar reviewed (commented):
Thanks for looking into this

kuhar commented:
Also FYI, turns out that some common select folds are illegal under the poison semantics: https://github.com/llvm/llvm-project/pull/84685. I don't think we have any such patterns in SPIR-V, but just wanted to mention this for some extra background on folding select.

inbelic committed:
review comments

- remove use of auto when type is unclear
- simplify the folding as we don't need to seperate the cases
- remove unneeded llvm namespacing
- use operand specific getters for readability

inbelic commented:
> Also FYI, turns out that some common select folds are illegal under the poison semantics

Neat

kuhar commented on mlir/lib/Dialect/SPIRV/IR/SPIRVCanonicalization.cpp:
```diff
@@ -801,16 +801,18 @@ OpFoldResult spirv::LogicalOrOp::fold(FoldAdaptor adaptor) {
-  SmallVector<ElementValueT, 4> elementResults;
+  SmallVector<Attribute, 4> elementResults;
   elementResults.reserve(condAttrs.getNumElements());
   for (size_t i = 0, e = condAttrs.getNumElements(); i < e;
```
nit
```suggestion
  size_t numElements = condAttrs.getNumElements();
  elementResults.reserve(numElements);
  for (size_t i = 0; i != numElements;
```

or if this works (iterators are sometimes weird and don't compose with all the utility functions):
```c++
auto elementResults = llvm::to_vector<4>(trueAttrs.getValues<Attribute>());
for (auto [result, cond, falseRes] : llvm::zip_equal(elementResults, condAttrs.getValues<BoolAttribute>(), falseAttr.getValues<Attribute>())) {
  if (cond.getValue())
    result = falseRes;
}
```

kuhar commented on mlir/lib/Dialect/SPIRV/IR/SPIRVCanonicalization.cpp:
```diff
@@ -872,7 +846,7 @@ OpFoldResult spirv::SelectOp::fold(FoldAdaptor adaptor) {
   // the scalar case. Hence, we are only required to consider the case of
-  // ElementsAttr in foldSelectOp.
+  // DenseElementsAttr in foldSelectOp.
   return foldSelectOp(operands);
```
Now that the logic is much simpler, I'd inline `foldSelectOp` here.

kuhar reviewed (commented):
A couple of small issues, otherwise looks great

inbelic committed:
review comments

- use abstractions to improve loop readability
- inline due to simplicity
